File: auth/src/main/java/com/ykb/app/cryptotrader/auth/component/filter/JwtFilter.java
package com.ykb.app.cryptotrader.auth.component.filter;

import com.ykb.app.cryptotrader.auth.service.EndpointService;
import com.ykb.app.cryptotrader.auth.service.JwtService;
import com.ykb.app.cryptotrader.auth.service.UserOperationsService;
import jakarta.servlet.FilterChain;
import jakarta.servlet.ServletException;
import jakarta.servlet.http.Cookie;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import org.springframework.http.HttpMethod;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.web.authentication.WebAuthenticationDetailsSource;
import org.springframework.stereotype.Component;
import org.springframework.web.filter.OncePerRequestFilter;

import java.io.IOException;
import java.util.Arrays;

@Component
public class JwtFilter extends OncePerRequestFilter {

    private final JwtService jwtService;
    private final UserOperationsService userService;
    private final EndpointService endpointService;

    public JwtFilter(JwtService jwtService, UserOperationsService userService, EndpointService endpointService) {
        this.jwtService=jwtService;
        this.userService=userService;
        this.endpointService = endpointService;
    }

    @Override
    protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response, FilterChain filterChain) throws ServletException, IOException {
        try {
            if ("OPTIONS".equalsIgnoreCase(request.getMethod())) {
                filterChain.doFilter(request, response);
                return;
            }

            if (requiresAuthentication(request)) {
                Authentication auth = attemptAuthentication(request);

                if (auth != null) {
                    SecurityContextHolder.getContext().setAuthentication(auth);
                } else {
                    SecurityContextHolder.clearContext();
                    response.sendError(HttpServletResponse.SC_UNAUTHORIZED, "Unauthorized");
                    return;
                }
            }

            filterChain.doFilter(request, response);
        } catch (RuntimeException ex) {
            SecurityContextHolder.clearContext();
            response.sendError(HttpServletResponse.SC_UNAUTHORIZED, "Invalid or expired token");
        }
    }

    private boolean requiresAuthentication(HttpServletRequest request) {
        var auth = SecurityContextHolder.getContext().getAuthentication();
        if(auth!=null && auth.isAuthenticated())
            return false;
        return endpointService.isAuthenticated(request.getRequestURI(), HttpMethod.valueOf(request.getMethod().toUpperCase()));
    }

    private Authentication attemptAuthentication(HttpServletRequest request) {
        var isView = endpointService.isView(request.getRequestURI(), HttpMethod.valueOf(request.getMethod().toUpperCase()));
        if(isView)
            return authenticateWithRefreshToken(request);

        return authenticateWithAccessToken(request);
    }

    private Authentication authenticateWithRefreshToken(HttpServletRequest request) {
        var jwt = getRefreshToken(request);
        String username = jwtService.authenticateRefreshToken(jwt);
        return authenticate(username, request);
    }

    private Authentication authenticateWithAccessToken(HttpServletRequest request) {
        var jwt = getAccessToken(request);
        String username = jwtService.authenticateAccessToken(jwt);
        return authenticate(username, request);
    }

    private Authentication authenticate(String username, HttpServletRequest request) {
        var user = userService.loadUserByUsername(username);
        var usernamePasswordAuthToken = new UsernamePasswordAuthenticationToken(user, null, user.getAuthorities());
        usernamePasswordAuthToken.setDetails(new WebAuthenticationDetailsSource().buildDetails(request));
        return usernamePasswordAuthToken;
    }

    private String getRefreshToken(HttpServletRequest request) {
        Cookie[] cookies = request.getCookies();
        if (cookies == null) return null;
        Cookie token = Arrays.stream(cookies)
                .filter(c -> "Authorization".equals(c.getName()))
                .findFirst()
                .orElse(null);
        return token != null ? token.getValue() : null;
    }

    private String getAccessToken(HttpServletRequest request) {
        String token = request.getHeader("Authorization");
        if(token!=null && token.startsWith("Bearer "))
            return token.substring(7);

        return null;
    }

}

File: library/auth/src/main/kotlin/com/ykb/app/cryptotrader/auth/filter/JwtFilter.kt
package com.ykb.app.cryptotrader.auth.filter

import org.springframework.http.HttpMethod
import com.ykb.app.cryptotrader.data.auth.model.User
import com.ykb.app.cryptotrader.auth.service.UserManager
import com.ykb.app.cryptotrader.auth.service.JwtManager
import com.ykb.app.cryptotrader.auth.service.UriSecurityManager
import com.ykb.app.cryptotrader.utils.logger.Logger
import jakarta.servlet.FilterChain
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.springframework.beans.factory.annotation.Value
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.security.web.AuthenticationEntryPoint
import org.springframework.security.web.authentication.WebAuthenticationDetailsSource
import org.springframework.security.web.savedrequest.RequestCache
import org.springframework.stereotype.Component

@Component
class JwtFilter(
    private val userManager: UserManager,
    private val jwtManager: JwtManager,
    private val uriSecurityManager: UriSecurityManager,
    authenticationEntryPoint: AuthenticationEntryPoint,
    requestCache: RequestCache,
    @Value("\${app.property.auth.url.login}") private val loginUri: String
) : HeaderFilter(authenticationEntryPoint, requestCache) {

    private val log: Logger = Logger.getLogger(JwtFilter::class)
    private val authCookieName = "Authorization"

    override fun doFilter(request: HttpServletRequest, response: HttpServletResponse, chain: FilterChain) {
        try {
            val authentication = SecurityContextHolder.getContext().authentication
            if(authentication != null && authentication.isAuthenticated) {
                chain.doFilter(request, response)
                return
            }

            val isLoginView = uriSecurityManager.isLoginView(request.requestURI, HttpMethod.valueOf(request.method.uppercase()))
            if(isLoginView) {
                chain.doFilter(request, response)
                return
            }

            val isView = uriSecurityManager.isView(request.requestURI, HttpMethod.valueOf(request.method))
            if(isView) {
                val jwt = getRefreshToken(request)
                if(!jwt.isNullOrBlank())
                    if(authenticateWithRefreshToken(jwt, request))
                        chain.doFilter(request, response)
                    else
                        response.sendRedirect(loginUri)
                else
                    response.sendRedirect(loginUri)
                return
            }

            val jwt = getAccessToken(request)
            if(!jwt.isNullOrBlank())
                authenticateWithAccessToken(jwt, request)
            chain.doFilter(request, response)
        } catch(ex: Exception) {
            handleException(request, response, chain, ex)
        }
    }

    private fun authenticateWithAccessToken(jwt: String, request: HttpServletRequest): Boolean {
        try {
            if (SecurityContextHolder.getContext().authentication == null && jwtManager.validateAccessToken(jwt)) {
                val username = jwtManager.getUsernameFromAccessToken(jwt)
                authenticateJwt(username!!, request)
                return true
            }
            return false
        } catch(ex: Exception) {
            return false
        }
    }

    private fun authenticateWithRefreshToken(jwt: String, request: HttpServletRequest): Boolean {
        try {
            if (SecurityContextHolder.getContext().authentication == null && jwtManager.validateRefreshToken(jwt)) {
                val username = jwtManager.getUsernameFromRefreshToken(jwt)
                authenticateJwt(username!!, request)
                return true
            }
            return false
        } catch(ex: Exception) {
            return false
        }
    }

    private fun authenticateJwt(username: String, request: HttpServletRequest) {
        val user: User = userManager.loadUserByUsername(username)
        val usernamePasswordAuthToken = UsernamePasswordAuthenticationToken(user, null, user.authorities)
        usernamePasswordAuthToken.details = WebAuthenticationDetailsSource().buildDetails(request)
        SecurityContextHolder.getContext().authentication = usernamePasswordAuthToken
    }

    private fun getRefreshToken(request: HttpServletRequest): String? {
        return try {
            request.cookies.first { cookie -> authCookieName==cookie.name }.value
        } catch (ex: Exception) {
            null
        }
    }

    private fun getAccessToken(request: HttpServletRequest): String? {
        var jwt: String? = null
        val requestTokenHeader = request.getHeader(authCookieName)
        if (requestTokenHeader != null && requestTokenHeader.startsWith("Bearer ")) {
            jwt = requestTokenHeader.substring(7)
        }
        return jwt
    }

}